bootloader: add riotboot minimal bootloader

kYc0o opened the pull request:
<!--
The RIOT community cares a lot about code quality.
Therefore, before describing what your contribution is about, we would like
you to make sure that your modifications are compliant with the RIOT
coding conventions, see https://github.com/RIOT-OS/RIOT/wiki/Coding-conventions.
-->
<!--
Put here the description of your contribution:
- describe which part(s) of RIOT is (are) involved
- if it's a bug fix, describe the bug that it solves and how it is solved
- you can also give more information to reviewers about how to test your changes
-->

### Contribution description
#### RIOT image header for riotboot bootloader
##### Overview
 
`riotboot` is the name of a minimal bootloader application and infrastructure.
 It consists of
  - the application "riotboot" in the bootloaders folder which serves as
    minimal bootloader
  - the module "riot_hdr" used to recognise RIOT firmware images
    which riotboot can boot.
  - the module "slot_util" used to manage the images (slots) with a
    RIOT header attached to them
  - a tool in dist/tools/riot_hdr for header generation.
  - several make targets to glue everything together

##### Concept
`riotboot` expects the flash to be separated in slots: slot 0 contains
 the bootloader and slot 1 a firmware image.

A RIOT image with a single slot looks like:
```
|------------------------------- FLASH -------------------------------------|
|- RIOTBOOT_SLOT0_SIZE -|-------- RIOTBOOT_FW_SLOT_SIZE (slot 1) -----------|
       (slot 0)         |---- RIOT_HDR_SIZE ------|
 ---------------------------------------------------------------------------
|       bootloader      | riot_hdr_t + filler (0) |     RIOT firmware       |
 ---------------------------------------------------------------------------
```
Please note that `RIOT_HDR_SIZE` depends on the architecture of the
MCU, since it needs to be aligned to either 256B or 512B. This is
fixed regardless of `sizeof(riot_hdr_t)`

The bootloader will, on reset, verify the checksum of the first slot, then
boot it. If the slot doesn't have a valid checksum, no image will be
booted and the bootloader will enter "while(1);".

The header contains

  - "RIOT" as magic number
  - an application version
  - a checksum

The bootloader "riotboot" only cares about checksum.


### Testing procedure

<!--
Details steps to test your contribution:
- which test/example to compile for which board and is there a 'test' command
- how to know that it was not working/available in master
- the expected success test output
-->
There's a test provided in tests/riotboot. This has only tested for now in:
  - [x] samr21-xpro
  - [x] iotlab-m3 (desk)

Please add above the boards that you have tested and worked.

### Issues/PRs references

<!--
Examples: Fixes #1234. See also #5678. Depends on PR #9876.

Please use keywords (e.g., fixes, resolve) with the links to the issues you
resolved, this way they will be automatically closed when your pull request
is merged. See https://help.github.com/articles/closing-issues-using-keywords/.
-->
#9342 
#9969

bergzand commented:
> Why did you cut out the slot selection logic?

From the description it looks like this is a minimal version that only supports a single firmware slot.

kaspar030 commented:
> From the description it looks like this is a minimal version that only supports a single firmware slot.

Yup, but my original commit contained like, ~15 lines more, but did support multiple slots.

IMO, going through all the bootloader hoops to verify a firmware's *metadata* hash seems a bit pointless.

kYc0o commented:
> Yup, but my original commit contained like, ~15 lines more, but did support multiple slots.

>IMO, going through all the bootloader hoops to verify a firmware's metadata hash seems a bit pointless.

I wanted to provide the minimal as possible and then build on top of it the extensions, like support for two firmware slots and flashing. I think I can re-add this logic since it doesn't hurt, but it will look like we're looking for inexistent slots (there are no support for second slot here).

Anyways, there are some current problems to be solved first, after solve them I can get back into this question.

emmanuelsearch commented:
@kYc0o any eta on when you will push the changes?

kYc0o commented:
> Anyways, there are some current problems to be solved first, after solve them I can get back into this question.

Marking WIP because of this.

> @kYc0o any eta on when you will push the changes?

Should be done by tomorrow at latest.

kYc0o commented:
Not WIP anymore.

I changed several conceptual things on this PR and remove some awful dependencies which were not needed.

Now, slot 0 represents the first firmware image, while the bootloader represents no slot. This change was to be aligned on other bootloaders (e.g. mcuboot) slot concepts, and actually having slot 0 as the bootloader wasn't really useful.

`RIOT_HDR_SIZE` is now removed and it's not anymore a dependency for the header generator, it's now rather passed as an argument.

The compilation of `genhdr` is now automatic, but only depends on the existence of the binary, so any modification to the code of `riot_hdr` is not tracked. @cladmi do you have an idea to do it without importing all the files from `riot_hdr`?

Finally, I don't know how to get rid of the cppcheck error, in my system I don't have it, most likely because I have a more recent cppcheck version.

kYc0o commented:
Pushed some changes to get rid of the cppcheck error. Although, it was a false negative IMHO so we should update Murdock to avoid those.

emmanuelsearch commented:
@kYc0o first commit reads "cpu/cortexm_common: initial multislot support".
renamed?

kYc0o commented:
> @kYc0o first commit reads "cpu/cortexm_common: initial multislot support".
renamed?

Will do.

kYc0o commented:
Done. Commits anyways would need reordering and more squashing in order to have working commits.

cladmi commented on dist/tools/riot_hdr/common.h:
```diff
@@ -0,0 +1,24 @@
+#define COMMON_H
+
+#include <unistd.h>
+#include <sys/types.h>
```
For `size_t` including `stddef.h` is enough.

kYc0o replied:
Will change.

cladmi commented on dist/tools/riot_hdr/genhdr.c:
```diff
@@ -0,0 +1,113 @@
+
+const char generate_usage[] = "genhdr generate <IMG_BIN> <APP_VER> <START_ADDR> <HDR_LEN> <outfile|->";
+
+int genhdr(int argc, char *argv[])
```
Or even better, the parsing could be done in `main.c` here it does not make sense to have the usage not defined in `main`.

kYc0o replied:
I'll give a try with the arguments on main, however in the future is meant to have more second arguments which differ on its usage. Check [here](https://github.com/RIOT-OS/RIOT/pull/9969/files#diff-6ba7583906bf62cc2dede299b3d0d1daR1).

cladmi replied:
Yes but then only the part after `generate` should be given to the `genhdr` parser and `usage` should do:

```printf("Usage genhdr generate %s", generate_usage);```
With `generate_usage` == `"<IMG_BIN> <APP_VER> <START_ADDR> <HDR_LEN> <outfile|->"`.

If not you have a hard written `"generate"` in your code which is duplicated.

cladmi commented on cpu/cortexm_common/include/cpu.h:
```diff
@@ -189,6 +189,44 @@ static inline void cortexm_isr_end(void)
+ *
+ * @param[in]   image_address   address in flash of other image
+ */
+static inline void cpu_jump_to_image(uint32_t image_address)
```
The argument is `image_address` so should be declared as an address type. As there is currently no reason to limit to `32bit` or `cortexm_common` in the `cpu_jump_to_image` I would declare it as `void *`.

kYc0o replied:
> As there is currently no reason to limit to 32bit or cortexm_common in the cpu_jump_to_image I would declare it as void *.

Yes there is: it doesn't (and would never) exist a cortex-m cpu which is not 32bit. Moreover, only certain cortex-m (already ifdef'd) support the relocation of the vector table (VTOR). Thus this code is not meant to be executed by any other architecture.

For more information: http://infocenter.arm.com/help/index.jsp?topic=/com.arm.doc.dui0552a/Ciheijba.html however they measure the offset in words instead of bytes.

danpetry replied:
Seems like this function should be called `jump_to_address`? Since it's being passed an address, not an image number

cladmi commented on bootloaders/riotboot/README.md:
```diff
@@ -0,0 +1,26 @@
+the version of the RIOT firmware `APP_VER`.
+This bootloader verifies the checksum of the image which is located
+at an offset (`ROM_OFFSET`) with respect to  the `ROM_START_ADDR`
+defined by the CPU, just after the space allocated for riotboot.
```
This documentation needs uptading. We are not verifying the checksum of the `image` put the metadata right ?
Also a bit of the same confusion between `image` and `slot` and `metadata`.

kYc0o replied:
Yes, I will try to read carefully everything and be consistent on the naming, since I agree it's confusing.

cladmi commented on bootloaders/riotboot/Makefile:
```diff
@@ -0,0 +1,26 @@
+DISABLE_MODULE += auto_init
+
+# Shared interface to read RIOT header
+USEMODULE += riot_hdr
```
Should come with `slot_util` directly.

kYc0o replied:
Agreed.

cladmi commented on bootloaders/riotboot/Makefile:
```diff
@@ -0,0 +1,26 @@
+BOARD ?= samr21-xpro
+
+# Select the boards that were tested and work with riotboot
+BOARD_WHITELIST += samr21-xpro iotlab-m3
```
I would rather put is as `FEATURES_REQUIRED += riotboot`.

kYc0o replied:
Agreed.

cladmi commented on dist/tools/riot_hdr/main.c:
```diff
@@ -0,0 +1,35 @@
+#include <string.h>
+
+int genhdr(int argc, char *argv[]);
+extern const char generate_usage[];
```
If these are shared they should go in a header instead of hacking this here without function documentation.

kYc0o replied:
`genhdr` is only used by main if the command `generate` is used.
`generate_usage` is only used in `genhdr`.

They're not shared but I can consider moving them to a .h, I just really need to see the befits.

cladmi commented on sys/include/riot_hdr.h:
```diff
@@ -0,0 +1,129 @@
+ * @ingroup     sys
+ * @{
+ *
+ * # RIOT image header for riotboot bootloader
```
This should not mention `riotboot` or explain how to compile it as the dependency is in the other order.
It would be really visible by having commits in the right order.

kYc0o replied:
I think this needs to be documented somewhere and riot_hdr doesn't make sense without riotboot using it here. True, riot_hdr can be used by other bootloaders but for now the usage is here.

I agree the commits need reorder as I said before, however I think this is a separated issue, as long as the usage of riot_hdr is showcased somewhere.

cladmi commented on sys/include/riot_hdr.h:
```diff
@@ -0,0 +1,129 @@
+ * A RIOT image with a single slot looks like:
+ *
+ * |------------------------------- FLASH -------------------------------------|
+ * |----- RIOTBOOT_LEN ----|----------- RIOTBOOT_SLOT_SIZE (slot 0) -----------|
```
It misses a possible existing bootloader.

cladmi commented on sys/include/riot_hdr.h:
```diff
@@ -0,0 +1,129 @@
+ * |----- RIOTBOOT_LEN ----|----------- RIOTBOOT_SLOT_SIZE (slot 0) -----------|
+ *                         |--- RIOTBOOT_HDR_LEN ----|
+ *  ---------------------------------------------------------------------------
+ * |       bootloader      | riot_hdr_t + filler (0) |     RIOT firmware       |
```
This should be `riotbootloader` to not mix with an existing one.

cladmi commented on sys/include/riot_hdr.h:
```diff
@@ -0,0 +1,129 @@
+ *
+ * - "RIOT" as magic number
+ * - an application version
+ * - a checksum
```
The header checksum

cladmi commented on sys/include/slot_util.h:
```diff
@@ -0,0 +1,88 @@
+/**
+ * @brief   Storage for addresses of the configured slots
+ */
+extern const uint32_t slot_util_slots[];
```
This is a storage for `addresses` so should be a pointer and not `uint32_t`.
Maybe even a `riot_hdr_t *` from its usage.

cladmi commented on sys/riot_hdr/riot_hdr.c:
```diff
@@ -0,0 +1,63 @@
+
+uint32_t riot_hdr_checksum(riot_hdr_t *riot_hdr)
+{
+    return fletcher32((uint16_t *)riot_hdr, RIOT_HDR_CHECKSUM_LEN / 2);
```
Could be `RIOT_HDR_CHECKSUM_LEN / sizeof(uint16_t)`

cladmi commented on sys/riot_hdr/riot_hdr.c:
```diff
@@ -0,0 +1,63 @@
+#include <string.h>
+
+#ifdef RIOT_VERSION
+#include "cpu.h"
```
Why is `cpu.h` required here ?

cladmi commented on sys/slot_util/Makefile:
```diff
@@ -0,0 +1,3 @@
+CFLAGS += -DRIOTBOOT_LEN=$(RIOTBOOT_LEN)
```
This variable should not be `RIOTBOOT` as it is  for`SLOT_UTIL`. See also the global comment.

cladmi commented on sys/slot_util/slot_util.c:
```diff
@@ -0,0 +1,139 @@
+ * Take into account that CPU_FLASH_BASE represents the starting
+ * address of the bootloader, thus we skip it.
+ */
+const uint32_t slot_util_slots[] = {
```
Same here as the documentation, this should not be `uint32_t` but a pointer type.

cladmi replied:
It could even simply be defined as `riot_hdr_t *` as it is what it is supposed to be.

cladmi commented on sys/slot_util/slot_util.c:
```diff
@@ -0,0 +1,139 @@
+ * address of the bootloader, thus we skip it.
+ */
+const uint32_t slot_util_slots[] = {
+    CPU_FLASH_BASE + RIOTBOOT_LEN,   /* First slot address -> firmware image */
```
`RIOTBOOT_LEN` see global comment, it does not take into account `ROM_OFFSET`.

cladmi commented on cpu/cortexm_common/include/cpu.h:
```diff
@@ -189,6 +189,44 @@ static inline void cortexm_isr_end(void)
+#if defined(CPU_ARCH_CORTEX_M0PLUS) || defined(CPU_ARCH_CORTEX_M3) || \
+    defined(CPU_ARCH_CORTEX_M4) || defined(CPU_ARCH_CORTEX_M4F) || \
+    defined(CPU_ARCH_CORTEX_M7)
+static inline uint32_t cpu_get_image_baseaddr(void)
```
Also an address defined as `uint32_t`.

kYc0o replied:
I think the representation of the address as an `uint32_t` eases quite a lot the usage of the function, taking into account that the vendor headers define addresses as unsigned integer constants. Changing everything into pointer makes the coding more complex and one can be easily lost casting those pointers.

cladmi commented on sys/slot_util/slot_util.c:
```diff
@@ -0,0 +1,139 @@
+
+int slot_util_current_slot(void)
+{
+    uint32_t base_addr = cpu_get_image_baseaddr();
```
More `current_image_address` than `base_addr`.

cladmi commented on sys/slot_util/slot_util.c:
```diff
@@ -0,0 +1,139 @@
+
+void slot_util_jump(unsigned slot)
+{
+    _assert_slot(slot);
```
This assert is done by `slot_util_get_hdr` already.

cladmi commented on sys/slot_util/slot_util.c:
```diff
@@ -0,0 +1,139 @@
+{
+    _assert_slot(slot);
+
+    riot_hdr_t *riot_hdr = (riot_hdr_t*)slot_util_slots[slot];
```
This is `slot_util_get_hdr` with the `_assert_slot` included.

cladmi commented on sys/slot_util/slot_util.c:
```diff
@@ -0,0 +1,139 @@
+    return 1;
+}
+
+unsigned slot_util_get_image_startaddr(unsigned slot)
```
Type is not even coherent here with the header.

kYc0o replied:
Leftover, however I wonder why Murdock didn't complain.

cladmi commented on sys/include/slot_util.h:
```diff
@@ -0,0 +1,88 @@
+ *
+ * @param[in]   slot    slot nr to jump to
+ */
+void slot_util_jump(unsigned slot);
```
I think it would make sense to declare it with `NORETURN`.

kYc0o replied:
Which are the (good) consequences of doing that? I don't really have a strong opinion.

cladmi commented on sys/include/slot_util.h:
```diff
@@ -0,0 +1,88 @@
+ *
+ * @returns address of first byte of @p slot
+ */
+uint32_t slot_util_get_image_startaddr(unsigned slot);
```
`uint32_t` is not an `address` type.

kYc0o replied:
As explained before, keeping the return values as the come from the origin eases a lot its handling in the code. I'd prefer to keep this representation as is, even though you're right they're addresses and a pointer type would be the right one.

cladmi commented on sys/include/riot_hdr.h:
```diff
@@ -0,0 +1,129 @@
+/**
+ *  @brief Number of header bytes that get checksummed
+ */
+#define RIOT_HDR_CHECKSUM_LEN      (12)
```
Should this be public ?

kYc0o replied:
No, I can move it to the .c file.

cladmi commented on sys/riot_hdr/riot_hdr.c:
```diff
@@ -0,0 +1,63 @@
+
+int riot_hdr_validate(riot_hdr_t *riot_hdr)
+{
+    if (memcmp(riot_hdr, "RIOT", 4)) {
```
This "RIOT" should not be a magic in here but declared as as public macro. It is also already used in two files.

kYc0o replied:
Agreed.

cladmi commented on sys/include/riot_hdr.h:
```diff
@@ -0,0 +1,129 @@
+ * @brief Structure to store image header - All members are little endian
+ * @{
+ */
+typedef struct {
```
What storage type endianness ?

kYc0o replied:
The same as its members.

bergzand replied:
What was the design decision by always having these members as little endian instead of host endian?

cladmi replied:
The endianness should be deterministic because the board reads it but it is generated by your computer.

bergzand replied:
I agree that the endianess should be deterministic.

Either we have to generate a struct on the host with the right (device) endianess or add little endian reads to the bootloader.

cladmi replied:
I asked to use `le32toh` when reading it but @kYc0o said it would make things more complicated… I found it more complicated without.

bergzand replied:
I don't really mind either solution, but IMHO it is not something we can simply ignore.

cladmi replied:
I was in favor of splitting `riot_hdr` in its own PR with its own test to be able to correctly review these things. I asked for it IRL but yeah.

cladmi commented on makefiles/boot/riotboot.mk:
```diff
@@ -0,0 +1,110 @@
+$(BINDIR)/$(APPLICATION)-slot0.elf: ROM_OFFSET=$(SLOT0_OFFSET)
+
+# create binary target with RIOT header
+$(BINDIR)/$(APPLICATION)-slot0.riot.bin: %.riot.bin: %.hdr %.bin
```
I would like to have this `$(BINDIR)/$(APPLICATION)-slot0.riot.bin` in a variable on top.

kYc0o replied:
Can do it.

cladmi commented on makefiles/boot/riotboot.mk:
```diff
@@ -0,0 +1,110 @@
+
+# slot 0 targets
+SLOT0_OFFSET := $$(($(RIOTBOOT_LEN) + $(RIOTBOOT_HDR_LEN)))
+$(BINDIR)/$(APPLICATION)-slot%.elf: FW_ROM_LEN=$(RIOTBOOT_SLOT_LEN - $(RIOTBOOT_HDR_LEN))
```
Same confusion here, this does not generates the `slot` but the `slot.firmware`. 

I would rename them to

`-slot%.firmware.bin` (or `fw.bin`), `-slot%.hdr.bin` and `slot%.bin`.

kYc0o replied:
> this does not generates the slot but the slot.firmware

I don't fully understand this. Maybe it's because I messed up with the definitions, but a slot should definitely be composed of header + firmware.

cladmi commented on Makefile.dep:
```diff
@@ -787,6 +787,14 @@ ifneq (,$(filter periph_gpio_irq,$(USEMODULE)))
 endif
 
+ifneq (,$(RIOTBOOT_LEN))
+  FEATURES_PROVIDED += riotboot
```
This is wrong, boards (boards common) should say `FEATURES_PROVIDED += riotboot`. It would remove the need to whitelist them.

Also, should it not be somehow a `slot` feature more than `riotboot` ? or both maybe.

kaspar030 replied:
The idea here is that if a board defined a RIOTBOOT_LEN, the board specific riotboot variables are present, thus the board supports the riotboot feature. If RIOTBOOT_LEN is not set, there's no riotboot for the board.

cladmi replied:
Currently it's defined in `cortexm_common/Makefile.include` which could be contested but there is not changes request on it yet. And using `cortexm_common` is not enough to support it. You need to use `cortexm_common/ldscript/cortexm.ld`, define the `ROM_ADDR`, … variables and have a cpu_init that can be called multiple times. The current configuration also currently does not support having an integrated bootloader or using `nordic_softdevice_ble`.

Also, it's a definition that would go basically in a `Makefile.include`, which is not included when parsing dependencies by `info-boards-supported` (and should not be). It is working right now when doing make as we parse them in a wrong order https://github.com/RIOT-OS/RIOT/issues/9913.

This shows me that `FEATURES_REQUIRED += riotboot` is missing in `tests/riotboot` or as a dependency of `slot_util` maybe.

```
git diff .
diff --git a/tests/riotboot/Makefile b/tests/riotboot/Makefile
index 3e49c4f0b..d0bcfcadb 100644
--- a/tests/riotboot/Makefile
+++ b/tests/riotboot/Makefile
@@ -10,6 +10,8 @@ BOARD_WHITELIST += samr21-xpro iotlab-m3
 USEMODULE += slot_util
 USEMODULE += riot_hdr

+FEATURES_REQUIRED += riotboot
+
 # Comment this out to disable code in RIOT that does safety checking
 # which is not needed in a production environment but helps in the
 # development process:

$ make info-boards-supported

```

kYc0o commented on sys/slot_util/slot_util.c:
```diff
@@ -0,0 +1,139 @@
+}
+
+#else /* BOARD_NATIVE */
+const unsigned slot_util_num_slots = 1;
```
It's a leftover, in the past slot_util and riot_hdr were the same. With current separation it's maybe not needed, will check.

kYc0o commented on tests/riotboot/Makefile:
```diff
@@ -0,0 +1,23 @@
+
+# Include modules to test the bootloader
+USEMODULE += slot_util
+USEMODULE += riot_hdr
```
Agree, will do.

kYc0o commented on dist/tools/riot_hdr/genhdr.c:
```diff
@@ -0,0 +1,113 @@
+    }
+
+    /* ensure the buffer and header have 0's */
+    memset(&riot_hdr, '\0', sizeof(riot_hdr_t));
```
It's based on a recommendation for the usage of `memset()`, which apparently converts it anyways into a `unsigned char`.

For reference https://stackoverflow.com/questions/23810453/why-use-0-instead-of-0-for-memset

kYc0o commented on dist/tools/riot_hdr/genhdr.c:
```diff
@@ -0,0 +1,113 @@
+ */
+#define HDR_ALIGN              (256)
+
+const char generate_usage[] = "genhdr generate <IMG_BIN> <APP_VER> <START_ADDR> <HDR_LEN> <outfile|->";
```
Style, both IMHO are ok so it's up to the developer, I don't have a strong opinion on this.

cladmi replied:
By using `char *` it is automatically defined a as const. And here it is actually just a string to print and not supposed to represent any memory chunk.

cladmi replied:
Hmm it can save the `pointer` as it is defined as a symbol, I did not think about that.

kYc0o commented on dist/tools/riot_hdr/genhdr.c:
```diff
@@ -0,0 +1,113 @@
+/**
+ *  @brief Alignment required to set VTOR in Cortex-M0+/3/4/7
+ */
+#define HDR_ALIGN              (256)
```
Actually it's always aligned to 256. It depends directly on the amount of interrupts decalred in the vector table. More info: http://infocenter.arm.com/help/index.jsp?topic=/com.arm.doc.dui0552a/Ciheijba.html

cladmi replied:
This tool is only called from the build system that already handles this.
Also nothing here, for the moment, is cortexm specific so no need to add it.

kYc0o replied:
Unfortunately the build system doesn't take care of any check avoiding the user to provide non-aligned header reserved space, besides a comment next to the default value. I prefer to keep it at the code level to ensure consistent headers.

kYc0o commented on dist/tools/riot_hdr/common.h:
```diff
@@ -0,0 +1,24 @@
+ extern "C" {
+#endif
+
+int to_file(const char *filename, void *buf, size_t len);
```
I'd like to keep it in common since more functions will come. See #9969.

I'll document it but I'm not sure documentation for dist/tools is created.

cladmi replied:
Generated to `docs.riot-os.org` or not, as a developer I am reading it.

From #9969 actually only the file management functions will be kept.

The `sha256` thing should be done by `riot_hdr_validate` or this file is doing more than just handling `riot_hdr`.
Or a `riot_hdr_validate_with_signature` if the header signature is not re-checked by the bootloader in `riot_hdr_validate`. But in that case, I would see the `OTA` mechanism being merged with the bootloader handling.

kYc0o commented:
> There is a mix between RIOTBOOT_LEN and RIOTBOOT_SLOT0_OFFSET as the first one does not take into account ROM_OFFSET that could come from an existing bootloader.

I agree that I need to be more consistent with the names, I'll make some changes and you can verify again. However, I don't want to deal with `ROM_OFFSET` right now, maybe I can make some changes to prepare a bit but I see it only as a nice to have.

kYc0o commented:
> For sys/slot_util to make it ready for adding another slot, the exported variables could be SLOTS_ROM_OFFSET, SLOT_LEN (and NUM_SLOT in the next one) as it would not require additional variables later.

I'll do something about this, I'm also preparing the multislot PR so I'll see how it integrates everything together.

>Currently SLOT0_OFFSET has the same mistake that I made before and should be made clearer in the doc/naming.
>
>A slot is composed of
>```
>Metadata  | firmware
>```
>
>So SLOT0_OFFSET should not include metadata. There should be a specific SLOT0_FIRMWARE_OFFSET.

I think this can be solved with a single variable. There's no need IMHO of any knowledge about the firmware part, but only where the beginning of the slot is.

> I would like to have tests/riotboot as a real test that would be executed by murdock by default.
We would need to currently hack the HEXFILE, IMAGE_FILE and override FLASHFILE to make it work.
So adapt it to do make all flash test.

This will also take some time, if you can propose something I'd gladly integrate it here.

kYc0o commented:
@jia200x I think we can wait for the hard feature freeze, but I'd say there are chances to get this in.

kYc0o commented:
@cladmi I think I addressed all of your comments, maybe I need to squash some fixups but let me know if it's reviewable as it is.

emmanuelsearch commented:
I tried on a stmf3discovery (adding `FEATURES_PROVIDED += riotboot` in its Makefile.features) but it fails to flash. I get the below error.
What needs to be done exactly for this board (or additional similar boards) to be supported?

```
### Flashing Target ###
Open On-Chip Debugger 0.10.0+dev-00167-g29cfe9c5 (2017-07-13-13:24)
Licensed under GNU GPL v2
For bug reports, read
	http://openocd.org/doc/doxygen/bugs.html
hla_swd
adapter speed: 1000 kHz
adapter_nsrst_delay: 100
Info : The selected transport took over low-level target control. The results might differ compared to plain JTAG/SWD
none separate
srst_only separate srst_nogate srst_open_drain connect_deassert_srst
Info : Unable to match requested speed 1000 kHz, using 950 kHz
Info : Unable to match requested speed 1000 kHz, using 950 kHz
Info : clock speed 950 kHz
Info : STLINK v2 JTAG v16 API v2 SWIM v0 VID 0x0483 PID 0x3748
Info : using stlink api v2
Info : Target voltage: 2.905039
Info : stm32f3x.cpu: hardware has 6 breakpoints, 4 watchpoints
    TargetName         Type       Endian TapName            State       
--  ------------------ ---------- ------ ------------------ ------------
 0* stm32f3x.cpu       hla_target little stm32f3x.cpu       halted
adapter speed: 1000 kHz
target halted due to debug-request, current mode: Thread 
xPSR: 0x01000000 pc: 0x08000368 msp: 0x20000200
auto erase enabled
Info : device id = 0x10036422
Info : flash size = 256kbytes
target halted due to breakpoint, current mode: Thread 
xPSR: 0x61000000 pc: 0x2000003a msp: 0x20000200
wrote 4096 bytes from file /Users/emmanuel/Documents/work/CODING/RIOT/bootloaders/riotboot/bin/stm32f3discovery/riotboot.elf in 0.329455s (12.141 KiB/s)
target halted due to breakpoint, current mode: Thread 
xPSR: 0x61000000 pc: 0x2000002e msp: 0x20000200
target halted due to breakpoint, current mode: Thread 
xPSR: 0x61000000 pc: 0x2000002e msp: 0x20000200
verified 2640 bytes in 0.150084s (17.178 KiB/s)
adapter speed: 1000 kHz
shutdown command invoked
Done flashing
/Users/emmanuel/Documents/work/CODING/RIOT/dist/tools/openocd/openocd.sh flash
### Flashing Target ###
sed: 1: "/^{.*}$/ {s/\} /\}\n/g; ...": extra characters at the end of p command
Binfile detected, adding ROM base address: 
/Users/emmanuel/Documents/work/CODING/RIOT/dist/tools/openocd/openocd.sh: line 207: 0x1000 + : syntax error: operand expected (error token is " ")
Flashing with IMAGE_OFFSET: 
Open On-Chip Debugger 0.10.0+dev-00167-g29cfe9c5 (2017-07-13-13:24)
Licensed under GNU GPL v2
For bug reports, read
	http://openocd.org/doc/doxygen/bugs.html
hla_swd
adapter speed: 1000 kHz
adapter_nsrst_delay: 100
Info : The selected transport took over low-level target control. The results might differ compared to plain JTAG/SWD
none separate
srst_only separate srst_nogate srst_open_drain connect_deassert_srst
Info : Unable to match requested speed 1000 kHz, using 950 kHz
Info : Unable to match requested speed 1000 kHz, using 950 kHz
Info : clock speed 950 kHz
Info : STLINK v2 JTAG v16 API v2 SWIM v0 VID 0x0483 PID 0x3748
Info : using stlink api v2
Info : Target voltage: 2.905039
Info : stm32f3x.cpu: hardware has 6 breakpoints, 4 watchpoints
    TargetName         Type       Endian TapName            State       
--  ------------------ ---------- ------ ------------------ ------------
 0* stm32f3x.cpu       hla_target little stm32f3x.cpu       running
adapter speed: 1000 kHz
target halted due to debug-request, current mode: Thread 
xPSR: 0x01000000 pc: 0x08000368 msp: 0x20000200
auto erase enabled
Error: Invalid command argument
image.base_address option value ('bin') is not valid

make: *** [riotboot/flash-slot0] Error 1
```

fedepell commented:
Tried on saml21-xpro: works good out of the box!

kYc0o commented:
@emmanuelsearch your problem is related to the `sed` command which differs from the `gnu-sed` (a.k.a. `gsed`) which is not available by default on OS X machines (which I know is your case).

@cladmi proposes #10165 to try to solve this, please try to review such PR. Otherwise installing `gsed` might be your last resource (`brew install gsed`).

kYc0o commented:
I added a last squash commit to provide some flexibility in case of we want a multislot support on which we can define different sized slots.

kYc0o commented:
@fedepell thanks a lot for your contribution! It's very useful, we will review it there and give feedback.

BTW, do you care reviewing this PR? It seems you're more or less familiar already with the design, plus you have also a current use case, so any feedback from you is very appreciated.

fedepell commented:
@kYc0o : ok, I'll happily try to review it tomorrow as today I'm OOO! Let me know in case any changes are needed for the PRs I submitted.

fedepell commented on sys/slot_util/slot_util.c:
```diff
@@ -0,0 +1,84 @@
+};
+
+/* Calculate the number of slots */
+const unsigned slot_util_num_slots = sizeof(slot_util_slots) / sizeof(uint32_t);
```
Formally shouldn't it be:
` const unsigned slot_util_num_slots = sizeof(slot_util_slots) / sizeof(riot_hdr_t *); `

cladmi replied:
+1

It is also common to use directly a variable item so here `slot_util_slots[0]`.

`#define ARRAY_LEN(array) (sizeof(array)/sizeof(array[0]))`

kYc0o replied:
It was a leftover, I fixed it.

fedepell commented on boards/samr21-xpro/Makefile.features:
```diff
@@ -9,6 +9,9 @@ FEATURES_PROVIDED += periph_spi
 FEATURES_PROVIDED += periph_uart
 
+# Put other features for this board (in alphabetical order)
+FEATURES_PROVIDED += riotboot
```
As I had it tested do you want to add it also to saml21-xpro board?

kYc0o replied:
Yes, I can do that myself.

kYc0o replied:
Done.

fedepell commented on makefiles/boot/riotboot.mk:
```diff
@@ -0,0 +1,120 @@
+# binary with both bootloader and RIOT firmware with header
+BOOTLOADER_BIN = $(RIOTBASE)/riotboot/bin/$(BOARD)
+$(BOOTLOADER_BIN)/riotboot.extended.bin: $(BOOTLOADER_BIN)/riotboot.bin
+	cp $^ $@.tmp
```
Use $(Q) consistently all around?

cladmi replied:
You can also directly use  `$(Q)cp $^ $@` here as the file can be in an intermediate state within the receipe.
`gcc` also takes a few cycles to write all bytes in an elf file.

kYc0o replied:
Done.

cladmi commented on makefiles/boot/riotboot.mk:
```diff
@@ -0,0 +1,120 @@
+
+# By default, slot 0 is found just after RIOTBOOT_LEN. It might
+# be overridden to add more offset as needed.
+export SLOT0_OFFSET ?= $(RIOTBOOT_LEN)
```
The `CPU_FLASH_BASE` is handled by both the linker script and the `slot_utils`.

Also when flashing a binary, `edbg` already includes the base address and requires `offset` and I made openocd behave the same way.

cladmi commented on sys/include/riot_hdr.h:
```diff
@@ -0,0 +1,86 @@
+ * @{
+ */
+typedef struct {
+    uint32_t magic_number;         /**< header magic_number (always "RIOT")    */
```
+1 I asked to precise the endianess of all fields.

kYc0o replied:
I really don't get what do you mean by this. Should I put an attribute to each field to express the endianness? Otherwise I don't know how to make explicit the endianness of a type or variable in C.

cladmi replied:
Sorry I missed it was written one line before the github shown section.

kYc0o commented on dist/tools/riot_hdr/main.c:
```diff
@@ -0,0 +1,36 @@
+    char *usage = "genhdr generate [args]";
+
+    if (argc < 2) {
+        goto usage;
```
Actually, I think the "upcoming" function won't really come. Historically, this tool was meant to fill the security fields of the header, which won't come anymore (security features are provided by other means), so maybe we can just drop it as @fedepell suggests.

kYc0o replied:
According to @bergzand he will take care of this in an upcoming PR.

fedepell commented on makefiles/boot/riotboot.mk:
```diff
@@ -0,0 +1,120 @@
+	$(Q)/usr/bin/env -i \
+		QUIET=$(QUIET) \
+		PATH=$(PATH) \
+			make --no-print-directory -C $(HEADER_TOOL_DIR) all
```
In general, in my experience, using $(MAKE) is preferable because $(MAKE) could have also flags near apart of the reference to the binary itself so you carry them along. So it could be that $(MAKE) is "make -x -y -z" and you are passing it forward.
Try to give a look also here: 
https://www.gnu.org/software/make/manual/html_node/MAKE-Variable.html#MAKE-Variable

kYc0o replied:
Agreed, I changed by `$(MAKE)`.

kYc0o commented on sys/include/riot_hdr.h:
```diff
@@ -0,0 +1,86 @@
+ * @brief  Validate image header
+ *
+ * @param[in] riot_hdr  ptr to image header
+ *
```
Fixed here and in the other documents. Strange that the doccheck didn't catch them.

cladmi replied:
Still not updated here.

cladmi replied:
Oh now its showing on my side… Github may still be lagging for some changes…

kYc0o commented on sys/riot_hdr/riot_hdr.c:
```diff
@@ -0,0 +1,67 @@
+
+void riot_hdr_print(const riot_hdr_t *riot_hdr)
+{
+    printf("Image magic_number: 0x%08x\n", (unsigned)riot_hdr->magic_number);
```
Sorry, it was a leftover. I uniformed everything as `0x%08x`.

jia200x commented:
is this going to make it for the Release? :)

kYc0o commented:
I really hope so! @cladmi ?

kYc0o commented:
I made the requirements explicit in the doc, I hope it helps to identify potential supported boards and the required changes to make them work with riotboot.

kYc0o commented on bootloaders/riotboot/README.md:
```diff
@@ -0,0 +1,72 @@
+  - Being able to execute startup code at least twice (`board_init()`)
+  - Being able to flash binary files, if integration with the build
+    system is required for flashing
+
```
Yes you're right, I'll change quickly.

kYc0o commented on bootloaders/riotboot/README.md:
```diff
@@ -0,0 +1,72 @@
+e.g. `BOARD=samr21-xpro APP_VER=$(date +%s) make -C examples/hello-world riotboot/flash`
+
+The command compiles both the hello-world example and riotboot,
+generates the header and attaches it at the beginning of the example
```
Done.

emmanuelsearch replied:
@kYc0o actually without spaces `FEATURES_REQUIRED+=riotboot` else it fails.
And how about adding `term` at the end of the command?

kYc0o replied:
Oh yeah, I just took it from the makefile so didn't mind the spaces, I'll correct it.

basilfx commented on Makefile.include:
```diff
@@ -705,6 +705,9 @@ CFLAGS += -include '$(RIOTBUILD_CONFIG_HEADER_C)'
 # include mcuboot support
 include $(RIOTMAKE)/mcuboot.mk
 
+# include bootlaoders support
```
Small typo: bootlaoders.

emmanuelsearch commented on sys/riot_hdr/riot_hdr.c:
```diff
@@ -0,0 +1,67 @@
+    printf("Image magic_number: 0x%08x\n", (unsigned)riot_hdr->magic_number);
+    printf("Image Version: 0x%08x\n", (unsigned)riot_hdr->version);
+    printf("Image start address: 0x%08x\n", (unsigned)riot_hdr->start_addr);
+    printf("Image chksum: 0x%08x\n", (unsigned)riot_hdr->chksum);
```
@kYc0o Based on what I understand from the `riot_hdr_t` strcut, this checksum is not the image checksum , but of the checksum of the header (metadata) itself. Am I getting this wrong? Anyhow: there seems to be a contradiction here. Needs clarification.

Typically, the Flash that is checksummed should include not only the metadata but also the image.
But maybe we can take this on in a subsequent PR?

kYc0o replied:
True, I can take care of the rewording, but as discussed below, we only aim to have a checksum of the header.

emmanuelsearch replied:
@kYc0o ok take care of rewording then

emmanuelsearch commented on sys/riot_hdr/riot_hdr.c:
```diff
@@ -0,0 +1,67 @@
+    if (res) {
+        LOG_INFO("%s: riot_hdr checksum invalid\n", __func__);
+    }
+
```
@kYc0o here should be a test checking the checksum of the Flash of the image, against what is in the metadata that should be check(summed) for validation should also contain the image.
(should we take that in a forthcoming PRs as suggested in my other comment?)

emmanuelsearch commented on sys/include/riot_hdr.h:
```diff
@@ -0,0 +1,88 @@
+    uint32_t version;              /**< Integer representing firmware version  */
+    uint32_t start_addr;           /**< Start address in flash                 */
+    uint32_t chksum;               /**< checksum of riot_hdr                   */
+} riot_hdr_t;
```
@kYc0o @cladmi as tentatively discussed with @cladmi a few days ago:
would it be simple to put the chksum in front of the struct, and add in the struct the size `image_size`
and then compute the checksum:
- starting from where the hdr + offset of sizeof(chksum)
- stopping at 256 + image_size ?

then `riot_hdr_checksum` could be easy to modify so that it checks the Flash of the image too (and not only the Flash of the hdr).

kaspar030 replied:
This would lead to increased boot-time (also from deep sleep wakeups).
Also, checking the whole image on boot makes it impossible to use part of the slot for dynamic data, e.g., a file system or configuration area.

Also I don't see any practical advantage in having the bootloader verify the flash over and over again.

The safety (not security) concept for flashing a slot is as follows:

(0. crypto-verify header & image)
1. erase header block (so it's checksum is invalid)
2. write image
3. verify image hash
4. write header block (including it's own hash)

If followed, that procedure makes sure that the image after the header is the one that was supposed to be flashed, should the header be valid (have a correct checksum).

So for the price of having do wait a couple of miliseconds *at each each wakeup*, we'd check whether the flash got corrupted *after* it was already flashed & verified once.

So what does the bootloader do if the newest slot has a broken checksum? With the current logic, it would boot an old (and possibly insecure) image, if one is available. The sane thing would be to not boot at all in that case, leading to a dead device. Is that desirable for all applications? Anyhow, the problem scope just got a lot larger.

I was very happy to have arrived at a very minimal bootloader. I suggest we keep it that way. If any image want's to check its own integrity, let the image check itself.

bergzand replied:
While I don't see any problems with putting the checksum as first element of the struct, I agree with @kaspar030 here. The function of the checksum is to be able to distinguish a filled slot from an empty (or corrupt `riot_hdr_t`). Extending the function of the checksum leads to a more complex design at almost no extra guarantees.

If we want to verify an image, I think it would be better to add a separate (sha-based) hash to the first flash page, somewhere after the metadata and have a separate bootloader application that is able to check the image integrity.

kYc0o replied:
I agree with both @kaspar030 and @bergzand, this minimal header is just for booting purposes, and afaik we don't aim to provide security or even safety at this level, only a way to recognise RIOT images which meet the requirements of a consistent header.

emmanuelsearch replied:
@kYc0o @kaspar030 @bergzand ok so let's take on this discussion in follow-up PRs / enhancements.

bergzand replied:
I think it is also a good idea to keep the magic number as the first element of the struct. This way we never get into trouble if there are different type of headers marked with different magic numbers

cladmi replied:
Good catch indeed.

kYc0o commented:
Addressed last comments.

Commit history is getting large... is still reviewable?

emmanuelsearch commented:
@kYc0o @kaspar030 @bergzand @cladmi from your perspectives are we close to ACKing/squashing/merging, or not?

waehlisch commented on bootloaders/riotboot/README.md:
```diff
@@ -0,0 +1,73 @@
+RIOT firmware.
+
+A RIOT firmware in a single slot is composed by:
+
```
add "```" before and after figure to improve the layout.

cladmi commented:
> Commit history is getting large... is still reviewable?

It was one of the reason why I asked to split in smaller parts earlier ;) It is still 1k added lines with 4 parts.
Also it is a pain that we cannot mark one comment as resolved even if the code did not change :/

Maybe you could just already re-organize commits so that the fixups follow the correct commit.

cladmi commented on makefiles/boot/riotboot.mk:
```diff
@@ -0,0 +1,121 @@
+APP_VER ?= 0
+
+# Final target for slot 0 with riot_hdr
+SLOT0_RIOT_FIRMWARE = $(BINDIR_APP)-slot0.riot.bin
```
That should not be called `FIRMWARE` if it is the `slot`.
And I still do not get the `.riot.` in the name

cladmi commented on tests/riotboot/tests/01-run.py:
```diff
@@ -0,0 +1,31 @@
+    # Other info is hardware/app dependant so we just check basic compliance
+    child.expect("Image Version: 0x[0-9a-fA-F]{8}")
+    child.expect("Image start address: 0x[0-9a-fA-F]{8}")
+    child.expect("Image chksum: 0x[0-9a-fA-F]{8}")
```
This is outdated with the code

```suggestion
    child.expect("Header chksum: 0x[0-9a-fA-F]{8}")
```

cladmi commented on sys/include/riot_hdr.h:
```diff
@@ -0,0 +1,88 @@
+    uint32_t magic_number;         /**< header magic_number (always "RIOT")    */
+    uint32_t version;              /**< Integer representing firmware version  */
+    uint32_t start_addr;           /**< Start address in flash                 */
+    uint32_t chksum;               /**< checksum of riot_hdr                   */
```
Why not just `checksum` ? I noticed it from the `Header chksum: ` output.

cladmi commented on makefiles/boot/riotboot.mk:
```diff
@@ -0,0 +1,121 @@
+endif
+
+# Create combined binary booloader + RIOT firmware with header
+riotboot/combined-slot0: $(BINDIR_APP)-slot0-combined.bin
```
This filename is repeated 5 times and I will need to repeat it again a few times in the test Makefile to allow doing `make flash test` and making murdock use it.
It would be better to have a dedicated variable for it.

kYc0o replied:
I suggest `RIOTBOOT_COMBINED_BIN`. Does it work for you?

cladmi commented on Makefile.include:
```diff
@@ -705,6 +705,9 @@ CFLAGS += -include '$(RIOTBUILD_CONFIG_HEADER_C)'
 include $(RIOTMAKE)/mcuboot.mk
 
+# include bootloaders support
+include $(RIOTMAKE)/boot/riotboot.mk
```
This include should be before the `link` target definition.

Because when trying to overwrite one variable like `HEXFILE`, the value will already have been evaluated when declaring the `link` target.
And later, selecting the right file to flash should only be `FLASHFILE = $(NAMED_OF_THE_COMBINED_FILE)` before including `Makefile.include`.

In general the rules should be the last things defined even if in `Makefile.include` is not completely the case for now.

cladmi commented:
I updated the test and normally made murdock run it.

kYc0o commented:
Pushed latest changes that should address comments from @cladmi and @waehlisch.

Local test succeeded with samr21-xpro.

kYc0o commented:
The riot_hdr has been splitted from this PR due to its complexity. See #10253.

kYc0o commented:
Does someone care if I squash to integrate cleanly #10253 ?

fedepell commented on tests/riotboot/tests/01-run.py:
```diff
@@ -0,0 +1,31 @@
+
+    # Ask for current slot header info and checks for basic output integrity
+    child.sendline("curslothdr")
+    # Magic number is "RIOT", check for little and big endian
```
Agreed! @kYc0o if you want to fix or I can do after you rebased!

kYc0o replied:
Wait a bit for the rebase, I think some of your commits will also be squashed.

fedepell replied:
> Wait a bit for the rebase, I think some of your commits will also be squashed.

Sure, no problem!

kYc0o replied:
Done, @fedepell you may proceed.

kaspar030 commented:
> Does someone care if I squash to integrate cleanly #10253 ?

please do!

kYc0o commented:
For some reason my local git user and the github user desynch... any ideas?

miri64 commented:
> For some reason my local git user and the github user desynch... any ideas?

New computer and different address configured in `git config user.email` I'd say ;-).

miri64 commented:
Yepp:
```
$ git show --pretty=fuller
commit a9f2d7227ce8123067f372e6807c63f66b9a483f
Author:     Gaëtan Harter <gaetan.harter@fu-berlin.de>
AuthorDate: Tue Oct 23 18:25:56 2018 +0200
Commit:     Francisco Acosta <f.acosta.ext@xxxxxxxx.com>
CommitDate: Tue Dec 4 13:08:32 2018 +0100

    tests/riotboot: allow doing 'make flash test' directly and enable CI
    
    This 'hacks' that `make flash` works with the combined file by default.
    This should also make it work for murdock to enable it in CI.
```

miri64 commented:
If you actually want to use the new e-mail address, you can add it under https://github.com/settings/emails

kYc0o commented:
> New computer and different address configured in git config --global user.email I'd say ;-).

Strangely is not a new computer, I think I set up a different git environment by running some commands required to push to another repo.

kYc0o commented:
@fedepell thanks for the commits! They're now merged, I'll rebase them and add you as co-author.

kaspar030 commented:
> Well, I suppose I need to fix codacy's complaints about the style isn't it? I didn't know that it's now enforced.

They're not. You'll need to address the reviewers' comments. ;)

kYc0o commented:
Oh, true. Well, so far everything was addressed and Murdock is green. @cladmi @danpetry, everything ok?

kaspar030 commented:
Ok, I've pushed commits to make slot_util a part of riotboot, too.

kYc0o commented:
ACK for @kaspar030 changes. Should I (re)squash?

kaspar030 commented:
> ACK for @kaspar030 changes. Should I (re)squash?

From my perspective, yes. Some of the commit messages need updating (e.g., "slot_util" -> "riotboot_slot").

kaspar030 commented:
> From my perspective, yes.

We might want to give @cladmi @danpetry some time to look at the fixups.

kYc0o commented:
Agreed.

danpetry commented:
ACK on the fundamentals. Happy with test coverage, it compiles and runs, and the concept and structure works. Also I think that having `hdr` and `slot` as submodules of `riotboot` makes sense for now, and we could move them to their own modules if we start to see uses for them?

Here are some further comments. Non-blocking because I'd rather discuss these points in a dedicated issue, with the outcomes hopefully having a more global effect, rather than here.
- having implicit build system magic to make the aforementioned submodules happen, does "feel" cleaner. But I don't think that newer users of RIOT would immediately recognize that they exist as submodules. They would have to have some understanding of our build system. Personally I think it's a good design philosophy to make stuff idiot-proof, explicit, and obvious for users, as far as possible. Maybe at some point a targeted discussion would be good to align on what should be a module/submodule/etc and what shouldn't, and how they are implemented, so it's consistent across the OS
- Re the tests, I would still prefer that all API calls are tested explicitly by being called in the test application. (`riotboot_slot_get_hdr` isn't called in the test application, so e.g. someone could decide to remove this function and just include the code in `riotboot_slot_print_hdr`, break the API for some users, and the tests would still pass.) This is again something more of a "global test philosophy" point so won't block.

Re. further review, I'm off now until Monday and might not have a chance to look more at this today - but if the other parts of review aren't done by someone else by then I'll do them then.

danpetry commented on sys/slot_util/Makefile:
```diff
@@ -0,0 +1,3 @@
+CFLAGS += -DSLOT0_OFFSET=$(SLOT0_OFFSET)
```
I think this file and folder needs to be deleted, right?

kaspar030 replied:
yup, done, thanks!

cladmi commented:
Quick note, there is this on the way https://github.com/RIOT-OS/RIOT/pull/10539 that will break the `openocd` flashing hack.
It will need to be replaced by `riotboot/flash-combined-slot0: ELFFILE=$(RIOTBOOT_COMBINED_BIN)` when merged.

emmanuelsearch commented:
@cladmi you mean line 98-99 in `makefiles/boot/riotboot.mk` right? 
Could you push the fix now that #10539 is merged?

cladmi commented:
Yep I will.

cladmi commented:
I also included the `openocd` commit that is in master so that the branch works even without rebasing it.

miri64 commented:
I asked @cladmi offline if it is ok to requeue this PR and he said yes, so I did (kind of want to get some PRs in today)

kYc0o commented:
Let me know when I must squash.

kYc0o commented:
Murdock is ok, so it's up to the reviewers .

cladmi commented:
While working on https://github.com/RIOT-OS/RIOT/pull/10494 I suspected that building in docker would also not be supported here.

```
BOARD=iotlab-m3 BUILD_IN_DOCKER=1 DOCKER="sudo docker" make -C tests/riotboot
make: Entering directory '/home/harter/work/git/RIOT/tests/riotboot'
compiling /home/harter/work/git/RIOT/dist/tools/riotboot_gen_hdr/bin/genhdr...
make: Nothing to be done for 'all'.
make: *** No rule to make target '/home/harter/work/git/RIOT/tests/riotboot/bin/iotlab-m3/tests_riotboot-slot0.bin', needed by '/home/harter/work/git/RIOT/tests/riotboot/bin/iotlab-m3/tests_riotboot-slot0.hdr'.  Stop.
make: Leaving directory '/home/harter/work/git/RIOT/tests/riotboot'
```

Also some things should be generated in docker and not outside, I will try to see what should be done to handle both cases.

cladmi commented:
I found a nice way that I should test more and polish:

After having `DOCKER_BINDIR` defined I can use the following patch.
There is still a choice on which file to wrap, to say which target is correctly supported from command line. I would say the phony targets should work, and could be enough, so wrap the first file target.
Also, right now I do not export the configuration variables but I should.

``` diff
diff --git a/makefiles/boot/riotboot.mk b/makefiles/boot/riotboot.mk
index ab0c6e63f..d628f4027 100644
--- a/makefiles/boot/riotboot.mk
+++ b/makefiles/boot/riotboot.mk
@@ -40,9 +40,14 @@ $(BINDIR_APP)-slot0.elf: FW_ROM_LEN=$$((SLOT0_LEN - $(RIOTBOOT_HDR_LEN)))
 $(BINDIR_APP)-slot0.elf: ROM_OFFSET=$(SLOT0_IMAGE_OFFSET)
 
 # Create binary target with RIOT header
+ifneq (1,$(BUILD_IN_DOCKER))
 $(SLOT0_RIOT_BIN): %.riot.bin: %.hdr %.bin
        @echo "creating $@..."
        $(Q)cat $^ > $@
+else
+$(SLOT0_RIOT_BIN): ..in-docker-container
+$(SLOT0_RIOT_BIN): DOCKER_MAKECMDGOALS += $(patsubst $(BINDIR)/%,$(DOCKER_BINDIR)/%,$(SLOT0_RIOT_BIN))
+endif
 
 # Compile header tool if it doesn't exist, force its compilation in case
 # some files changed
@@ -89,8 +94,14 @@ endif
 # Create combined binary booloader + RIOT firmware with header
 RIOTBOOT_COMBINED_BIN = $(BINDIR_APP)-slot0-combined.bin
 riotboot/combined-slot0: $(RIOTBOOT_COMBINED_BIN)
+
+ifneq (1,$(BUILD_IN_DOCKER))
 $(RIOTBOOT_COMBINED_BIN): $(BOOTLOADER_BIN)/riotboot.extended.bin $(SLOT0_RIOT_BIN)
        $(Q)cat $^ > $@
+else
+$(RIOTBOOT_COMBINED_BIN): ..in-docker-container
+$(RIOTBOOT_COMBINED_BIN): DOCKER_MAKECMDGOALS += $(patsubst $(BINDIR)/%,$(DOCKER_BINDIR)/%,$(RIOTBOOT_COMBINED_BIN))
+endif
 
 # Flashing rule for edbg to flash combined binaries
 riotboot/flash-combined-slot0: HEXFILE=$(RIOTBOOT_COMBINED_BIN)
```

jcarrano commented on bootloaders/riotboot/Makefile:
```diff
@@ -0,0 +1,25 @@
+# limit riotboot bootloader size
+# TODO: Manage to set this variable for boards which already embed a
+# bootloader, currently it will be overwritten
+ROM_LEN := $(RIOTBOOT_LEN)
```
Why the `:=`?

kYc0o replied:
This variable needs to be evaluated at the moment of its parsing, since the value is taken as a reference for further calculations. Otherwise the value changes and the results are non-consistent.

jcarrano commented on bootloaders/riotboot/Makefile:
```diff
@@ -0,0 +1,25 @@
+FEATURES_REQUIRED += riotboot
+
+# Disable unused modules
+CFLAGS += -DNDEBUG -DLOG_LEVEL=LOG_NONE
```
Shouldn't this be conditional on DEBUG or DEVELHELP or something like that?

kYc0o replied:
In this case this is used intentionally to reduce at its maximum the size of the bootloader. Otherwise instances of the debugging module might be activated and thus making use of the libc, which increases the size of the bootloader by several kilobytes.

Activating them under other conditions implies to increase the size of the bootloader, which is not done automatically, and requires several modifications to the current approach.

jcarrano commented on cpu/cortexm_common/Makefile.include:
```diff
@@ -24,3 +24,11 @@ TOOLCHAINS_SUPPORTED = gnu llvm
+RIOTBOOT_LEN ?= 0x1000
+# Take the whole flash minus RIOTBOOT_LEN
+SLOT0_LEN ?= $(shell printf "0x%x" $$(($(ROM_LEN:%K=%*1024)-$(RIOTBOOT_LEN))))
+SLOT0_LEN := $(SLOT0_LEN)
```
What's the purpose of this assignment?

kYc0o replied:
The calculations done in the line above are only performed at the end of the make parsing, but the value should be consistent across the whole process. This ensures the value is calculated and preserved as is.

jcarrano commented on dist/tools/riotboot_gen_hdr/Makefile:
```diff
@@ -0,0 +1,39 @@
+RIOTBASE       := ../../..
```
Please remove all the instances of `:=`

kYc0o replied:
Does it really hurts?

jcarrano replied:
Yes, because a few months from now, and nobody will know if the immediate assignments were necessary or not and if they can be safely removed.

And also (and this has nothing to do with the bootloader) it perpetuates the myth that `:=` is a sort of magical spell that one puts when things don't work.

jcarrano commented on dist/tools/riotboot_gen_hdr/Makefile:
```diff
@@ -0,0 +1,39 @@
+COMMON_HDR     := common.h
+
+RIOT_HDR_SRC := \
+	$(RIOTBASE)/sys/checksum/fletcher32.c \
```
Is it correct to peek into the contents of a module? I feel this violates module encapsulation.

kYc0o replied:
The compilation of the tools is a complete separated process, actually in a complete different environment. This line only makes (re)use of the C file, which has no dependencies so it's just like copy/pase it.

jcarrano commented on makefiles/boot/riotboot.mk:
```diff
@@ -0,0 +1,122 @@
+
+# Generate RIOT header and keep the original binary file
+# It must be always regenerated in case of any changes, so FORCE
+.PRECIOUS: %.bin
```
Why the .PRECIOUS? Ideally, if you need the .bin for some rule you should state that explicitly.

kYc0o replied:
It's more for debugging purposes. The intermediate bin files are not really needed, but provide information about its generation which helps to know if they are correct. On the other hand, they might also be distributed separately. This is mostly an optional thing which does't affect neither the compilation process (speed) or the size of the build (their size is just a few bytes).

jcarrano replied:
Ok, undesrtood. Again, maybe it's good to document it so that we know that it is there for debugging only. Even then, I think that if the bin is wanted it should be explicitly made a dependency of riotboot or something like that.

jcarrano commented on makefiles/boot/riotboot.mk:
```diff
@@ -0,0 +1,122 @@
+			$(MAKE) --no-print-directory -C $(HEADER_TOOL_DIR) all
+
+# Generate RIOT header and keep the original binary file
+# It must be always regenerated in case of any changes, so FORCE
```
In case of changes to what? Changes to the tool and the binary are already being considered because of the prerequisites.

kYc0o replied:
For some reason the tool compilation was not re-triggered after modifying those files. I (we, I did this together with @cladmi) didn't investigate further.

jcarrano commented on makefiles/boot/riotboot.mk:
```diff
@@ -0,0 +1,122 @@
+BOOTLOADER_BIN = $(RIOTBOOT_DIR)/bin/$(BOARD)
+$(BOOTLOADER_BIN)/riotboot.extended.bin: $(BOOTLOADER_BIN)/riotboot.bin
+	$(Q)cp $^ $@.tmp
+	$(Q)truncate -s $$(($(RIOTBOOT_LEN))) $@.tmp
```
you can get the same result without resorting to a temporary file:
```
cat $^ /dev/zero | dd bs=1 count=$(RIOTBOOT_LEN) of=$@
```

kYc0o replied:
I think your suggestion assumes the existence of `/dev/zero` and `dd`, the proposed solution just `truncate`. What are the benefits of your proposition besides saving some bytes?

jcarrano replied:
Not leaving a tmp file behind if it fails. Not a big deal, but it's cleaner.

jcarrano commented on makefiles/boot/riotboot.mk:
```diff
@@ -0,0 +1,122 @@
+# Create binary target with RIOT header
+$(SLOT0_RIOT_BIN): %.riot.bin: %.hdr %.bin
+	@echo "creating $@..."
+	$(Q)cat $^ > $@
```
Protect this with lazysponge to prevent partial output.

kYc0o replied:
What's lazysponge?

jcarrano replied:
It's a tool by @cladmi (in dist/tools). A sort of remake of [sponge from more utils](https://linux.die.net/man/1/sponge)

It buffers stdin and then spits it out to a file all at once.

kYc0o replied:
I still don't get what's the benefit of having it here. You mean that something else could write to the stdin and corrupt the binary?

jcarrano commented on makefiles/boot/riotboot.mk:
```diff
@@ -0,0 +1,122 @@
+RIOTBOOT_HDR_LEN ?= 0x100
+
+# Export variables for 'riotboot_slot'
+export SLOT0_LEN
```
What are these exports for?

kYc0o replied:
It's used in the process of calculation of `SLOT0_LEN` which might be variable (helps also for the multislot case).

danpetry commented on dist/tools/riotboot_gen_hdr/genhdr.c:
```diff
@@ -0,0 +1,114 @@
+    if (!to_file(argv[5], hdr_buf, hdr_len)) {
+        fprintf(stderr, "Error: cannot write output\n");
+        free(hdr_buf);
+        return(1);
```
should be `return 1;` like the other `return`s?

kYc0o replied:
It's exactly the same.

danpetry replied:
Stylistic issue, not a functional one

danpetry replied:
Merged, thanks

kYc0o commented on dist/tools/riotboot_gen_hdr/common.c:
```diff
@@ -0,0 +1,39 @@
+{
+    int fd;
+
+    if (strcmp("-", filename)) {
```
it's a kind of wildcard to accept any name of file provided through the command line.

danpetry replied:
ok thanks!

kYc0o commented on sys/riotboot/slot.c:
```diff
@@ -0,0 +1,90 @@
+ * space reserved to the bootloader.
+ */
+const riotboot_hdr_t * const riotboot_slots[] = {
+    (riotboot_hdr_t*)(CPU_FLASH_BASE + SLOT0_OFFSET),   /* First slot address -> firmware image */
```
This was discussed offline during one of the task force meetings (IIRC you were there). This is to keep compatibility with the single slot case (the one proposed in this PR) while the existence of a second slot might be expected (the case of multislot). Decoupling the number of slots from the proposed code might lead to extra complexity for a case we are not sure will come (e.g. more than two slots), so my bet is that even for a 3 or 4 slots case, the code still maintainable.

There might be other (more) clever ways to deal with this, but I didn't have (neither have now) the time to put into it and thus keeping this PR open for several more months (or even a year).

danpetry replied:
> Decoupling the number of slots from the proposed code might lead to extra complexity for a case we are not sure will come
This makes sense. We can always change stuff later if needed

kYc0o commented on sys/include/riotboot/slot.h:
```diff
@@ -0,0 +1,109 @@
+ *
+ * @returns 0 if ok.
+ */
+static inline int riotboot_slot_validate(unsigned slot)
```
The goal is to avoid any explicit dependency of the modules using `riotboot_slot` from `riotboot_hdr`. Encapsulating here this function avoids to include `riotboot_hdr.h`.

The other function only uses code belonging to this module, so no need to encapsulate it in any way.

danpetry replied:
> Encapsulating here this function avoids to include riotboot_hdr.h.

You can keep `#include riotboot/hdr.h` in `slot.h`, move the function definitions to `slot.c`, and keep only the declarations in `slot.h`. The user wouldn't need to explicitly include `riotboot_hdr.h` in that case either. And you wouldn't have function definitions in header files, which is generally bad practice.

kYc0o commented on sys/riotboot/slot.c:
```diff
@@ -0,0 +1,90 @@
+
+uint32_t riotboot_slot_get_image_startaddr(unsigned slot)
+{
+    return riotboot_slot_get_hdr(slot)->start_addr;
```
I'm actually unsure that `hdr` can be a NULL pointer. And not because it will always be there, but because we're returning an adress from flash, which I'm not sure it can be `NULL`.

IIRC I think this line is a debug leftover so it can be removed. Which other function makes use of it?

danpetry replied:
just `riotboot_slot_dump_addrs`, sorry for the slightly confusing phrasing.

kYc0o replied:
I just checked the removal of the `hdr == NULL` and it still works properly.

danpetry replied:
Merged, thanks

kYc0o commented:
> I'm still not sure about the `SLOT0_LEN`, `ROM_LEN` calculation stuff.

Can you find at least one point of failure?

danpetry commented:
Thanks for addressing comments. The tests still pass. ACK from me.

jcarrano commented:
@kYc0o 
> Can you find at least one point of failure?
Not of failure, but those mechanisms in make that depend on order tend to be be fragile. In any case, I wrote that as a comment and not as a change request because I realize it's not an issue of this PR, and there's not much you can do about it.

emmanuelsearch commented:
@cladmi must the docker aspect you mention be part of this PR (or can it be a subsequent PR)?

emmanuelsearch commented:
@jcarrano what is blocking (if anything) in your comments, from your point of view?

emmanuelsearch commented:
ping @cladmi

cladmi commented:
@emmanuelsearch issue is also present for `mcuboot` integration so not a new issue.
I will try to fix it for both for the release.

danpetry commented:
Please squash and rebase. @kaspar030 do you need any more changes before merging?

kaspar030 commented:
IMO this is good to go. Please squash!

kaspar030 commented:
I squashed as @kYc0o is currently offline. As I co-authored most of the code, we'll need more than my ACK. Anyhow, I think we should get this in ASAP. We can always iterate.

emmanuelsearch commented:
@kaspar030 we have another ACK from @danpetry so we're good if Murdock is green.

kYc0o commented:
😃 a big thanks to all the contributors and reviewers to make this PR in!

waehlisch commented:
@jcarrano 
> It's easier to fix it myself in a PR later.

We don't want forget this: Can you create issues or the PR timely.
